File: hydrograph.engine/hydrograph.engine.cascading/src/main/java/hydrograph/engine/execution/tracking/JobInfo.java
/*******************************************************************************
 * Copyright 2016 Capital One Services, LLC and Bitwise, Inc.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/
package hydrograph.engine.execution.tracking;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import cascading.flow.FlowNode;
import cascading.flow.FlowStep;
import cascading.flow.planner.Scope;
import cascading.flow.planner.graph.ElementGraph;
import cascading.pipe.Pipe;
import cascading.stats.CascadingStats;
import cascading.stats.FlowNodeStats;
import cascading.stats.FlowStepStats;

public class JobInfo {

	private final String COUNTER_GROUP = "com.hydrograph.customgroup";
	private Map<String, ComponentInfo> componentInfoMap = new HashMap<>();
	private Map<String, Pipe> componentPipeMap;
	private Map<String, String> pipeComponentMap;
	private Map<String, List<String>> componentSocketMap;
	private Map<String, List<String>> componentAndPreviousMap;
	private List<String> AllPipes;
	private static List<String> listOfFilterComponent;

	/**
	 * Processes the CascadingStats to generate the component Statistics
	 * 
	 * @param cascadingStats
	 * @throws ElementGraphNotFoundException
	 */
	public synchronized void storeComponentStats(CascadingStats<?> cascadingStats)
			throws ElementGraphNotFoundException {
		checkAndCreateMaps();
		generateStats(cascadingStats);
	}

	private void checkAndCreateMaps() {
		if (componentPipeMap == null) {
			componentPipeMap = ComponentPipeMapping.getComponentToPipeMapping();
			pipeComponentMap = createReverseMap(componentPipeMap);
			componentSocketMap = ComponentPipeMapping.getComponentSocketMap();
			componentAndPreviousMap = ComponentPipeMapping.getComponentAndPreviousMap();
			AllPipes = new ArrayList<String>();
			listOfFilterComponent = ComponentPipeMapping.getListOfFilterComponent();
		}
	}

	private void generateStats(CascadingStats<?> cascadingStats) throws ElementGraphNotFoundException {
		ElementGraph elementGraph = extractElementGraphFromCascadeStats(cascadingStats);
		for (Scope scope : elementGraph.edgeSet()) {
			if (!AllPipes.contains(scope.getName())) {
				AllPipes.add(scope.getName());
			}
		}
		for (Scope scope : elementGraph.edgeSet()) {
			String currentComponent_SocketId = getComponentFromPipe(scope.getName());
			if (currentComponent_SocketId != null) {
				String currentComponentId = getComponentIdFromComponentSocketID(currentComponent_SocketId);
				if (!isComponentGeneratedFilter(currentComponentId)) {
					getPreviousComponentInfoIfScopeIsNotPresent(cascadingStats, currentComponentId);
					createComponentInfoForComponent(currentComponent_SocketId, cascadingStats);
					generateStatsForOutputComponent(currentComponent_SocketId, cascadingStats);
				}
			}
		}
	}

	private void generateStatsForOutputComponent(String currentComponent_SocketId, CascadingStats<?> cascadingStats) {
		String currentComponentId = getComponentIdFromComponentSocketID(currentComponent_SocketId);
		ComponentInfo currentComponentInfo = componentInfoMap.get(currentComponentId);
		List<String> previousComponents = new ArrayList<String>();
		Collection<List<String>> previousComponentLists = componentAndPreviousMap.values();
		for (List<String> previousComponentList : previousComponentLists) {
			previousComponents.addAll(previousComponentList);
		}
		if (!previousComponents.contains(currentComponent_SocketId)) {
			for (String previousGeneratedfilter : componentAndPreviousMap.get(currentComponentId)) {
				for (String previousComponent_SocketID : componentAndPreviousMap
						.get(getComponentIdFromComponentSocketID(previousGeneratedfilter))) {
					long recordCount = 0;
					String previousPipeName = componentPipeMap.get(previousComponent_SocketID).getName();
					recordCount = cascadingStats.getCounterValue(COUNTER_GROUP, previousPipeName);
					currentComponentInfo.setProcessedRecordCount("NoSocketId", recordCount);
				}
			}
		}

	}

	private boolean isComponentGeneratedFilter(String currentComponentId) {
		return listOfFilterComponent.contains(currentComponentId);
	}

	private void getPreviousComponentInfoIfScopeIsNotPresent(CascadingStats<?> cascadingStats,
			String currentComponentId) {
		if (componentAndPreviousMap.containsKey(currentComponentId)) {
			List<String> previousComponentId_SocketIds = componentAndPreviousMap.get(currentComponentId);
			for (String previousComponentId_SocketId : previousComponentId_SocketIds) {
				if (isComponentGeneratedFilter(getComponentIdFromComponentSocketID(
						getComponentFromPipe(componentPipeMap.get(previousComponentId_SocketId).getName())))) {
					List<String> prePreviousComponentId_SocketIds = componentAndPreviousMap
							.get(getComponentIdFromComponentSocketID(getComponentFromPipe(
									componentPipeMap.get(previousComponentId_SocketId).getName())));
					for (String prePreviousComponentId_SocketId : prePreviousComponentId_SocketIds) {
						String previousPipeName = componentPipeMap.get(prePreviousComponentId_SocketId).getName();
						if (!AllPipes.contains(previousPipeName) && previousPipeName != null) {
							createComponentInfoForComponent(prePreviousComponentId_SocketId, cascadingStats);
							getPreviousComponentInfoIfScopeIsNotPresent(cascadingStats,
									getComponentIdFromComponentSocketID(prePreviousComponentId_SocketId));
						}
					}
				}
			}
		}
	}

	private void createComponentInfoForComponent(String component_SocketId, CascadingStats<?> cascadingStats) {
		ComponentInfo componentInfo = null;
		String currentComponentId = getComponentIdFromComponentSocketID(component_SocketId);
		if (currentComponentId != null) {
			if (componentInfoMap.containsKey(currentComponentId)) {
				componentInfo = componentInfoMap.get(currentComponentId);
				componentInfo.setStatusPerSocketMap(getSocketIdFromComponentSocketID(component_SocketId),
						cascadingStats.getStatus());
			} else {
				componentInfo = new ComponentInfo();
				componentInfo.setComponentId(currentComponentId);
				for (String socketId : componentSocketMap.get(currentComponentId)) {
					componentInfo.setStatusPerSocketMap(socketId, cascadingStats.getStatus());
					componentInfo.setProcessedRecordCount(socketId, 0);
				}
			}
		}
		generateAndUpdateComponentRecordCount(cascadingStats, componentInfo, currentComponentId);
		componentInfoMap.put(currentComponentId, componentInfo);
		setStatus(componentInfo, cascadingStats);
	}

	private ElementGraph extractElementGraphFromCascadeStats(CascadingStats<?> cascadingStats)
			throws ElementGraphNotFoundException {
		ElementGraph elementGraph = null;
		if (cascadingStats instanceof FlowNodeStats) {
			FlowNodeStats flowNodeStats = (FlowNodeStats) cascadingStats;
			FlowNode flowNode = flowNodeStats.getFlowNode();
			elementGraph = flowNode.getElementGraph();
		} else if (cascadingStats instanceof FlowStepStats) {
			FlowStepStats flowStepStats = (FlowStepStats) cascadingStats;
			FlowStep<?> flowStep = flowStepStats.getFlowStep();
			elementGraph = flowStep.getElementGraph();
		} else {
			throw new ElementGraphNotFoundException(
					"Element Graph not found from FlowNodeStats/FlowStepStats while fetching stats from cascading");
		}
		return elementGraph;
	}

	private String getComponentIdFromComponentSocketID(String componentId_SocketId) {
		for (Entry<String, List<String>> component_Socketid : componentSocketMap.entrySet()) {
			String componentId = component_Socketid.getKey();
			for (String socketId : component_Socketid.getValue()) {
				if (componentId_SocketId.equals(componentId + "_" + socketId)) {
					return componentId;
				}
			}
		}
		return null;
	}

	private String getSocketIdFromComponentSocketID(String prevComponentId_socketid) {
		for (Entry<String, List<String>> component_Socketid : componentSocketMap.entrySet()) {
			String componentId = component_Socketid.getKey();
			for (String socketId : component_Socketid.getValue()) {
				if (prevComponentId_socketid.equals(componentId + "_" + socketId)) {
					return socketId;
				}
			}
		}
		return null;
	}

	private void generateAndUpdateComponentRecordCount(CascadingStats<?> cascadingStats, ComponentInfo componentInfo,
			String ComponentId) {
		for (String socketId : componentSocketMap.get(ComponentId)) {
			long recordCount = 0;
			for (String counter : cascadingStats.getCountersFor(COUNTER_GROUP)) {
				String componentIdANdSocketId = pipeComponentMap.get(counter);
				if ((getComponentIdFromComponentSocketID(componentIdANdSocketId) != null
						&& getComponentIdFromComponentSocketID(componentIdANdSocketId).equals(ComponentId))
						&& socketId.equals(getSocketIdFromComponentSocketID(componentIdANdSocketId))) {
					recordCount = cascadingStats.getCounterValue(COUNTER_GROUP, counter);
					componentInfo.setProcessedRecordCount(socketId, recordCount);
				}
			}
		}
	}

	private Map<String, String> createReverseMap(Map<String, Pipe> allMapOfPipes) {
		Map<String, String> pipeComponent = new HashMap<>();
		for (Map.Entry<String, Pipe> entry : allMapOfPipes.entrySet()) {
			pipeComponent.put(entry.getValue().getName(), entry.getKey());
		}
		return pipeComponent;
	}

	private String getComponentFromPipe(String pipeName) {
		for (Entry<String, Pipe> componentPipeSet : componentPipeMap.entrySet()) {
			if (componentPipeSet.getValue().getName().equals(pipeName)) {
				return componentPipeSet.getKey();
			}
		}
		return null;
	}

	private void setStatus(ComponentInfo componentInfo, CascadingStats<?> flowNodeStats) {
		Map<String, String> outSocketStats = componentInfo.getStatusPerSocketMap();
		List<String> listOfStatus = new ArrayList<String>();
		for (Entry<String, String> entry : outSocketStats.entrySet()) {
			listOfStatus.add(entry.getValue());
		}
		if (listOfStatus.contains("FAILED") || listOfStatus.contains("STOPPED"))
			componentInfo.setCurrentStatus("FAILED");
		else if (listOfStatus.contains("RUNNING"))
			componentInfo.setCurrentStatus("RUNNING");
		else if (listOfStatus.contains("PENDING") || listOfStatus.contains("STARTED")
				|| listOfStatus.contains("SUBMITTED")) {
			componentInfo.setCurrentStatus("PENDING");
		} else if (listOfStatus.contains("SUCCESSFUL")) {
			boolean isSuccessful = true;
			for (String status1 : listOfStatus) {
				if (!status1.equals("SUCCESSFUL"))
					isSuccessful = false;
			}
			if (isSuccessful)
				componentInfo.setCurrentStatus("SUCCESSFUL");
		}
	}

	/**
	 * Method returns the current status of all components
	 * 
	 * @return List of ComponentInfo
	 */
	public List<ComponentInfo> getstatus() {
		return new ArrayList<>(componentInfoMap.values());
	}

	/**
	 * ElementGraphNotFoundException
	 */
	public class ElementGraphNotFoundException extends Exception {

		private static final long serialVersionUID = 4691438024426481804L;

		public ElementGraphNotFoundException(String message) {
			super(message);
		}

	}
}

File: hydrograph.engine/hydrograph.engine.spark/src/main/scala/hydrograph/engine/spark/execution/tracking/JobInfo.scala
/*******************************************************************************
  * Copyright 2017 Capital One Services, LLC and Bitwise, Inc.
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * http://www.apache.org/licenses/LICENSE-2.0
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  *******************************************************************************/
package hydrograph.engine.spark.execution.tracking

import hydrograph.engine.execution.tracking
import hydrograph.engine.execution.tracking.ComponentInfo
import hydrograph.engine.spark.executiontracking.plugin.Component
import org.apache.spark.scheduler._

import scala.collection.JavaConverters._
import scala.collection.immutable.HashMap
import scala.collection.mutable
import scala.collection.mutable.ListBuffer

/**
  * Created by Aniketmo on 12/22/2016.
  */
class JobInfo(componentInfoMap: mutable.ListBuffer[Component]) {

  val componentInfoList = new java.util.ArrayList[tracking.ComponentInfo]
  val componentInfoHashSet = new mutable.HashSet[tracking.ComponentInfo]


  def storeComponentStatsForTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
    generateStatsForTaskEnd(taskEnd)
  }

  def storeComponentStatsForTaskStart(taskStart: SparkListenerTaskStart): Unit = {
    generateStatsForTaskStart(taskStart)
  }

  def storeComponentStatsForTaskGettingResult(taskGettingResult: SparkListenerTaskGettingResult): Unit = {
    generateStatsForTaskGettingResult(taskGettingResult)
  }

  def updateStatusOfComponentsOnStageSubmitted(stageSubmitted: SparkListenerStageSubmitted): Unit = {
    val Status = if(stageSubmitted.stageInfo.failureReason.equals(None))"SUCCESSFUL"
    else "FAILED"
    val listOfStatus : mutable.ListBuffer[String] = new ListBuffer[String]
    componentInfoList.asScala.filter(cInfo => cInfo.getStageId.
      equals(stageSubmitted.stageInfo.stageId)).
      foreach(compInfo=>{

        if(compInfo.getStatusPerSocketMap().values().iterator().hasNext){
          for(value <- compInfo.getStatusPerSocketMap().values().asScala){
            listOfStatus+=value
          }}

        if(listOfStatus.contains("FAILED")){
          for(key : String <- compInfo.getStatusPerSocketMap().keySet().asScala){
            compInfo.setProcessedRecordCount(key,-1)
          }
          compInfo.setCurrentStatus("FAILED")
        }
        else if (listOfStatus.contains("RUNNING")) {
          compInfo.setCurrentStatus("RUNNING")
        }
        else if (listOfStatus.contains("SUCCESSFUL")) {
          compInfo.setCurrentStatus("SUCCESSFUL")
        }

      })

  }


  def updateStatusOfComponents(stageCompleted: SparkListenerStageCompleted): Unit = {
    val Status = if(stageCompleted.stageInfo.failureReason.equals(None))"SUCCESSFUL"
    else "FAILED"
    val listOfStatus : mutable.ListBuffer[String] = new ListBuffer[String]
    componentInfoList.asScala.filter(cInfo => cInfo.getStageId.
      equals(stageCompleted.stageInfo.stageId)).
      foreach(compInfo=>{

        if(compInfo.getStatusPerSocketMap().values().iterator().hasNext){
          for(value <- compInfo.getStatusPerSocketMap().values().asScala){
            listOfStatus+=value
          }}

        if(listOfStatus.contains("FAILED")){
          for(key : String <- compInfo.getStatusPerSocketMap().keySet().asScala){
            compInfo.setProcessedRecordCount(key,-1)
          }
          compInfo.setCurrentStatus("FAILED")
        }
        else if (listOfStatus.contains("RUNNING")) {
          compInfo.setCurrentStatus("RUNNING")
        }
        else if (listOfStatus.contains("SUCCESSFUL")) {
          compInfo.setCurrentStatus("SUCCESSFUL")
        }

      })

    //logic for status of component whose compInfo is not generated
    /*componentInfoList.asScala.filter(cI=>cI.getStageId.equals(-1)).foreach(co=>{
      co.setCurrentStatus("SUCCESSFUL")
      for(key : String <- co.getStatusPerSocketMap().keySet().asScala){
        co.setStatusPerSocketMap(key,"SUCCESSFUL")
      }
    })*/

  }

  private def generateStatsForTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = {
    //    ElementGraph elementGraph = extractElementGraphFromCascadeStats(cascadingStats);
    val Status = if(taskEnd.taskInfo.status.equals("SUCCESS"))"SUCCESSFUL"
    else taskEnd.taskInfo.status
    //    println("Status in TaskEnd : "+Status)

    /*if(taskEnd.taskInfo.status.equals("FAILED"))
    componentInfoList.asScala.forall(a=>{if (!a.getCurrentStatus.equals("SUCCESSFUL"))
      a.setCurrentStatus("FAILED"); true})*/

    taskEnd.taskInfo.accumulables.foreach(f => {

      //      println("acumulator name : "+f.name.get + "accumulator value : "+ f.value.get)
      componentInfoMap.filter(c => c.newComponentId.equals(f.name.get)).foreach(component => {


        val alreadyPresentCompInfo  = componentInfoList.asScala
          .filter(compInfo=> compInfo.getComponentId.equals(component.compId))

        //        println("Reached after alreadyPresentCompInfo checking of taskEnd")
        if (alreadyPresentCompInfo.size > 0) {

          componentInfoList.asScala
            .filter(compInfo => {
              compInfo.getComponentId.equals(component.compId) && compInfo.getCurrentStatus() != null && !compInfo.getCurrentStatus().equals("FAILED")
            }).foreach(componentInfo => {
            componentInfo.setStageId(taskEnd.stageId)
            componentInfo.setComponentId(component.compId)
            componentInfo.setBatch(component.batch)
            componentInfo.setComponentName(component.compName)
//            f.value.foreach(x => println("PRINTING " +x))

            val tempMap = f.value.iterator.next().asInstanceOf[HashMap[Int, HashMap[Int, (Long, Int)]]]
            var recordCount = 0L
            for((stageId,hashmap)<-tempMap){
              if(stageId.equals(taskEnd.stageId)){
//                println("StageID "+stageId)
                for((intval,longintmap)<-hashmap){
//                  println("intval "+ intval)
                  recordCount = longintmap._1
//                    println("longValue "+ longintmap._1)
//                    println("integerVal "+ longintmap._2)
                }
              }
            }
            /*val itr = f.value.iterator.next().asInstanceOf[HashMap[Int, HashMap[Int, (Long, Int)]]].iterator
            if(itr.hasNext){
              if(itr.next()._1.equals(taskEnd.stageId)){
                val count = itr.next()._2.asInstanceOf[HashMap[Int, (Long, Int)]].iterator.next()._2.asInstanceOf[Tuple2[String,String]]._1
                println("FETCHED COUNT "+count)
              }
            }*/
            /*println("RECORD COUNT IN TASK END "+f.value.iterator.next().asInstanceOf[HashMap].iterator.next()._2.asInstanceOf[HashMap].iterator.next()._2.asInstanceOf[Tuple2]._1.asInstanceOf[Long])*/
            componentInfo.setProcessedRecordCount(component.outSocket, recordCount)
            componentInfo.setStatusPerSocketMap(component.outSocket, Status)
            if(Status.equals("FAILED")){
              componentInfo.setCurrentStatus("FAILED")
              componentInfo.setProcessedRecordCount(component.outSocket, -1)
            }
            else if(componentInfo.getCurrentStatus!=null && componentInfo.getCurrentStatus.equals("PENDING")){
              componentInfo.setCurrentStatus("RUNNING")
            }

            //            println("**************mark status as running of taskEnd******************")
            //           componentInfoHashSet.add(componentInfo)
          })

          //logic for status of component whose compInfo is not generated
          /*componentInfoList.asScala
            .filter(c=>c.getCurrentStatus.equals("PENDING")).foreach(ci=>
            {ci.setCurrentStatus("RUNNING")
              for(key : String <- ci.getStatusPerSocketMap().keySet().asScala){
                       ci.setStatusPerSocketMap(key,"RUNNING")
                   }
            })*/
        }
        /*else {
          val componentInfo = new ComponentInfo()
          componentInfo.setStageId(taskEnd.stageId)
          componentInfo.setComponentId(component.compId)
          componentInfo.setBatch(component.batch)
          componentInfo.setComponentName(component.compName)
          componentInfo.setProcessedRecordCount(component.outSocket, f.value.get.asInstanceOf[Long])
          componentInfo.setStatusPerSocketMap(component.outSocket, Status)
          componentInfo.setCurrentStatus("PENDING")
//          println("mark status as PENDING of taskEnd")
//          componentInfoHashSet.add(componentInfo)
          componentInfoList.add(componentInfo)
        }*/


      })
    })
  }

  private def generateStatsForTaskStart(taskStart: SparkListenerTaskStart): Unit = {
    //    ElementGraph elementGraph = extractElementGraphFromCascadeStats(cascadingStats);
    val Status = taskStart.taskInfo.status
    //    println("Status in TaskStart : "+Status)
    /*if(taskGettingResult.taskInfo.status.equals("SUCCESS"))"SUCCESSFUL"
    else taskStart.taskInfo.status*/

    /*if(taskStart.taskInfo.status.equals("FAILED"))
      componentInfoList.asScala.forall(a=>{if (!a.getCurrentStatus.equals("SUCCESSFUL"))
        a.setCurrentStatus("FAILED"); true})*/

    taskStart.taskInfo.accumulables.foreach(f => {
      //      println("acumulator name : "+f.name.get + "accumulator value : "+ f.value.get)
      componentInfoMap.filter(c => c.newComponentId.equals(f.name.get)).foreach(component => {


        val alreadyPresentCompInfo  = componentInfoList.asScala
          .filter(compInfo=> compInfo.getComponentId.equals(component.compId))

        //        println("Reached after alreadyPresentCompInfo checking of taskStart")
        if (alreadyPresentCompInfo.size > 0) {

          componentInfoList.asScala
            .filter(compInfo => {
              compInfo.getComponentId.equals(component.compId) && !compInfo.getCurrentStatus().equals("FAILED")
            }).foreach(componentInfo => {
            componentInfo.setStageId(taskStart.stageId)
            componentInfo.setComponentId(component.compId)
            componentInfo.setBatch(component.batch)
            componentInfo.setComponentName(component.compName)
            f.value.foreach(x => println("PRINTING " +x))

            val tempMap = f.value.iterator.next().asInstanceOf[HashMap[Int, HashMap[Int, (Long, Int)]]]
            var recordCount = 0L
            for((stageId,hashmap)<-tempMap){
              if(stageId.equals(taskStart.stageId)){
                println("StageID "+stageId)
                for((intval,longintmap)<-hashmap){
                  println("intval "+ intval)
                  recordCount = longintmap._1
                  println("longValue "+ longintmap._1)
                  println("integerVal "+ longintmap._2)
                }
              }
            }
            /*val itr = f.value.iterator.next().asInstanceOf[HashMap[Int, HashMap[Int, (Long, Int)]]].iterator
            if(itr.hasNext){
              if(itr.next()._1.equals(taskStart.stageId)){
                val count = itr.next()._2.asInstanceOf[HashMap[Int, (Long, Int)]].iterator.next()._2.asInstanceOf[Tuple2[String,String]]._1
                println("FETCHED COUNT "+count)
              }
            }*/
            /*println("RECORD COUNT IN TASK END "+f.value.iterator.next().asInstanceOf[HashMap].iterator.next()._2.asInstanceOf[HashMap].iterator.next()._2.asInstanceOf[Tuple2]._1.asInstanceOf[Long])*/
            componentInfo.setProcessedRecordCount(component.outSocket, recordCount)
            componentInfo.setStatusPerSocketMap(component.outSocket, Status)
            if(Status.equals("FAILED")){
              componentInfo.setCurrentStatus("FAILED")
              componentInfo.setProcessedRecordCount(component.outSocket, -1)
            }
            else if(componentInfo.getCurrentStatus!=null && componentInfo.getCurrentStatus.equals("PENDING")){
              componentInfo.setCurrentStatus("RUNNING")
            }
            //            println("mark status as running of taskStart")
            //           componentInfoHashSet.add(componentInfo)
          })
        }
        /*else {
          val componentInfo = new ComponentInfo()
          componentInfo.setStageId(taskStart.stageId)
          componentInfo.setComponentId(component.compId)
          componentInfo.setBatch(component.batch)
          componentInfo.setComponentName(component.compName)
          componentInfo.setProcessedRecordCount(component.outSocket, f.value.get.asInstanceOf[Long])
          componentInfo.setStatusPerSocketMap(component.outSocket, Status)
          componentInfo.setCurrentStatus("PENDING")
//          println("mark status as PENDING of taskStart")
          //          componentInfoHashSet.add(componentInfo)
          componentInfoList.add(componentInfo)
        }*/


      })
    })
  }

  //

  private def generateStatsForTaskGettingResult(taskStart: SparkListenerTaskGettingResult): Unit = {
    //    ElementGraph elementGraph = extractElementGraphFromCascadeStats(cascadingStats);
    /*val Status = taskStart.taskInfo.status
    println("Status in TaskStart : "+Status)
    /*if(taskGettingResult.taskInfo.status.equals("SUCCESS"))"SUCCESSFUL"
    else taskStart.taskInfo.status*/
    taskStart.taskInfo.accumulables.foreach(f => {
      //      println("acumulator name : "+f.name.get + "accumulator value : "+ f.value.get)
      componentInfoMap.filter(c => c.newComponentId.equals(f.name.get)).foreach(component => {


        val alreadyPresentCompInfo  = componentInfoList.asScala
          .filter(compInfo=> compInfo.getComponentId.equals(component.compId))

        //        println("Reached after alreadyPresentCompInfo checking of taskStart")
        if (alreadyPresentCompInfo.size > 0) {

          componentInfoList.asScala
            .filter(compInfo => {
              compInfo.getComponentId.equals(component.compId)
            }).foreach(componentInfo => {
//            componentInfo.setStageId(taskStart.stageId)
            componentInfo.setComponentId(component.compId)
            componentInfo.setBatch(component.batch)
            componentInfo.setComponentName(component.compName)
            componentInfo.setProcessedRecordCount(component.outSocket, f.value.get.asInstanceOf[Long])
            componentInfo.setStatusPerSocketMap(component.outSocket, Status)
            componentInfo.setCurrentStatus("RUNNING")
            //            println("mark status as running of taskStart")
            //           componentInfoHashSet.add(componentInfo)
          })
        }
       /* else {
          val componentInfo = new ComponentInfo()
//          componentInfo.setStageId(taskStart.stageId)
          componentInfo.setComponentId(component.compId)
          componentInfo.setBatch(component.batch)
          componentInfo.setComponentName(component.compName)
          componentInfo.setProcessedRecordCount(component.outSocket, f.value.get.asInstanceOf[Long])
          componentInfo.setStatusPerSocketMap(component.outSocket, Status)
          componentInfo.setCurrentStatus("PENDING")
          //          println("mark status as PENDING of taskStart")
          //          componentInfoHashSet.add(componentInfo)
          componentInfoList.add(componentInfo)
        }*/


      })
    })*/
  }


  def getStatus(): java.util.List[ComponentInfo] = {
    componentInfoList //componentInfoMap.values.toList
  }

  def createComponentInfos(): Unit ={

    componentInfoMap.foreach({comp=>
      val componentInfo = new ComponentInfo()
      componentInfo.setStageId(-1)
      componentInfo.setComponentId(comp.compId)
      componentInfo.setBatch(comp.batch)
      componentInfo.setComponentName(comp.compName)
      componentInfo.setProcessedRecordCount(comp.outSocket,0)
      componentInfo.setStatusPerSocketMap(comp.outSocket,"")
      componentInfo.setCurrentStatus("")
      //          println("mark status as PENDING of taskStart")
      //          componentInfoHashSet.add(componentInfo)
      componentInfoList.add(componentInfo)
    })
  }

  def getComponentInfoMap(): mutable.ListBuffer[Component] ={
    componentInfoMap
  }


}